src/metabase/automagic_dashboards/core.clj

```diff
@@ -0,0 +1,568 @@
+                                           :max-cardinality max_cardinality})
+                           (map #(assoc % :link (:link table)))))
+                (filter-tables tablespec (:tables context)))
+        (filter-fields {:fieldspec       tablespec
+                        :named           named
+                        :max-cardinality max_cardinality}
+                       (-> context :root-table :fields))))))
+
+(defn- make-binding
+  [context [identifier definition]]
+  (->> definition
+       (field-candidates context)
+       (map #(->> (merge % definition)
+                  vector
+                  (assoc definition :matches)
+                  (hash-map (name identifier))))))
+
+(def ^:private ^{:arglists '([definitions])} most-specific-definition
+  "Return the most specific defintion among `definitions`.
+   Specificity is determined based on:
+   1) how many ancestors `field_type` has (if field_type has a table prefix,
+      ancestors for both table and field are counted);
+   2) if there is a tie, how many additional filters (`named`, `max_cardinality`,
+      `links_to`, ...) are used;
+   3) if there is still a tie, `score`."
+  (comp last (partial sort-by (comp (fn [[_ definition]]
+                                      [(->> definition
+                                            :field_type
+                                            (map (comp count ancestors))
+                                            (reduce + ))
+                                       (count definition)
+                                       (:score definition)])
+                                    first))))
+
+(defn- bind-dimensions
+  "Bind fields to dimensions and resolve overloading.
+   Each field x aggregation pair will be bound to only one dimension. If multiple
+   dimension definitions match a single field, the field is bound to the specific
+   definition is used (see `most-specific-defintion` for details)."
+  [context dimensions]
+  (->> dimensions
+       (mapcat (comp (partial make-binding context) first))
+       (group-by (comp (juxt :id :aggregation) first :matches val first))
+       (map (comp most-specific-definition val))
+       (apply merge-with (fn [a b]
+                           (case (compare (:score a) (:score b))
+                             1  a
+                             0  (update a :matches concat (:matches b))
+                             -1 b))
+              {})))
+
+(defn- build-order-by
+  [dimensions metrics order-by]
+  (let [dimensions (set dimensions)]
+    (for [[identifier ordering] (map first order-by)]
+      [(if (= ordering "ascending")
+         :asc
+         :desc)
+       (if (dimensions identifier)
+         [:dimension identifier]
+         [:aggregate-field (u/index-of #{identifier} metrics)])])))
+
+(defn- build-query
+  ([context bindings filters metrics dimensions limit order_by]
+   (walk/postwalk
+    (fn [subform]
+      (if (rules/dimension-form? subform)
```

Comment:
Where are you expecting to find dimension-forms? Seems like it will be traversing a lot of stuff that couldn't possibly be a dimension form (i.e. `:type` and `:query` and `[:type :query]`)

Reply:
Anywhere in :query so it's just 2 keys, :type and :database which doesn't fell like that much.